### ISSUE TITLE
Warn when a generic constructor collapses to the same signature as a non-generic constructor for certain type arguments

### ISSUE THREAD

**Issue by Stuart88** [dotnet/roslyn#81473]
**Summary**

Requesting an analyzer diagnostic for cases where a generic constructor’s substituted signature becomes identical to a non-generic constructor, causing overload resolution to pick an unexpected constructor without warning.

___

For a generic class constructor that can accept its generic type as a parameter or a specific type parameter that matches the generic type, for example:

```csharp

class MyGenericClass<T>
{
    public MyGenericClass(T data) { }
    public MyGenericClass(string message) { }
}

var x = new MyGenericClass<string>("hello");

```

When `T` is `string`, both constructors become:

- `MyGenericClass(string data)`
- `MyGenericClass(string message)`

C# resolves the call to the non-generic overload, but the developer may expect the generic constructor to be used. There is no warning that the generic constructor’s signature collapses into another constructor’s signature for the chosen type argument.
 
Why this is a problem:

- Behaviour changes silently based on the substituted type.
- Easy for developers to overlook.
- Other similar ambiguities (e.g. hidden members) have diagnostics.

**Languages applicable:**

C#

**Code example that the analyzer should report:**

For the example `MyGenericClass<T>` above:

`var x = new MyGenericClass<string>("hello");`

This is a case where:

- A generic constructor’s substituted signature matches an existing constructor’s signature for the actual type argument, and
- Overload resolution will therefore pick a different constructor than the generic one.

**Comment by CyrusNajmabadi** [comment]
Unlikely to create our own warning for this.  But the analyzer framework exists so you can create your own analyzer for this exact purpose.  Come to discord if you'd like help with how to create such an analyzer.
